game-core/src/main/java/games/strategy/triplea/util/MovableUnitsFilter.java

```diff
@@ -87,81 +86,105 @@ public FilterOperationResult(
    */
-  public FilterOperationResult filterUnitsThatCanMove(
-      final Collection<Unit> units, final Map<Unit, Collection<Unit>> dependentUnits) {
-    final Collection<Unit> transportsToLoad =
-        getPossibleTransportsToLoad(units, dependentUnits, route);
+  public FilterOperationResult filterUnitsThatCanMove(final Collection<Unit> units) {
+    final List<Unit> best = getInitialUnitList(units);
 
-    List<Unit> best = getInitialUnitList(units);
+    final Collection<Unit> transportsToLoad = getPossibleTransportsToLoad(units);
     MoveValidationResultWithDependents lastResult =
-        validateMoveWithDependents(best, dependentUnits, transportsToLoad);
+        validateMoveWithDependents(best, transportsToLoad);
     final MoveValidationResult allUnitsResult = lastResult.getResult();
-
     if (!allUnitsResult.isMoveValid()) {
-      // if the player is invading only consider units that can invade
-      if (!nonCombat
-          && route.isUnload()
-          && Matches.isTerritoryEnemy(player, data).test(route.getEnd())) {
-        best = CollectionUtils.getMatches(best, Matches.unitCanInvade());
-        lastResult = validateMoveWithDependents(best, dependentUnits, transportsToLoad);
-      }
-      final boolean hasLandTransports =
-          TechAttachment.isMechanizedInfantry(player)
-              && best.stream().anyMatch(Matches.unitIsLandTransport());
-      final Predicate<Unit> isLandTransportable = Matches.unitIsLandTransportable();
-
-      while (!best.isEmpty() && !lastResult.getResult().isMoveValid()) {
-        final Unit firstSkipUnit = best.get(0);
-        int startIndex = 1;
-        // Check if we can skip more than one unit if they are equivalent (e.g. all units
-        // of the same type that have equivalent movement left). Don't do this if there are
-        // land transports (mech infantry), though.
-        while (startIndex < best.size()
-            && (!hasLandTransports || !isLandTransportable.test(best.get(startIndex)))
-            && unitsAreEquivalentWithSameMovementLeft(firstSkipUnit, best.get(startIndex))) {
-          startIndex++;
-        }
-        best = best.subList(startIndex, best.size());
-        lastResult = validateMoveWithDependents(best, dependentUnits, transportsToLoad);
-      }
+      lastResult = chooseSubsetOfUnitsThatCanMove(best, transportsToLoad, lastResult);
     }
 
     return new FilterOperationResult(allUnitsResult, lastResult);
   }
 
+  private MoveValidationResultWithDependents chooseSubsetOfUnitsThatCanMove(
+      final List<Unit> units,
+      final Collection<Unit> transportsToLoad,
+      final MoveValidationResultWithDependents initialResult) {
+    if (!transportsToLoad.isEmpty()) {
+      final List<Unit> allUnits = addMustMoveWith(units);
+      final Collection<Unit> loadedUnits =
+          TransportUtils.mapTransports(route, allUnits, transportsToLoad).keySet();
+      return validateMoveWithDependents(loadedUnits, transportsToLoad);
+    }
```

Comment:
This if case does make sense to me, but when is transportsToLoad empty?
And what does the rest of the method do in that case?
I really have no clue about the logic

Reply:
The rest of the logic is actually unchanged - refactored from the previous code.